noobaa_sa/s3_client.py

```diff
@@ -312,3 +307,83 @@ def delete_all_objects_in_bucket(self, bucket_name):
+
+    def generate_objects_for_upload(
+        self,
+        amount=1,
+        min_size="1M",
+        max_size="1M",
+        files_dir="",
+    ):
+        """
+        Generate random objects for upload purpose
+
+        Args:
+            amount (int): The number of objects to write
+            min_size(str): The minimum size of each object,
+                        specified in a format understood by the 'dd' command.
+            max_size(str): The maximum size of each object,
+                        specified in a format understood by the 'dd' command.
+            files_dir (str): A directory where the objects will be written
+                        locally. If not specified, a temporary directory will
+                        be used.
+
+        Returns:
+            list: A list of the names of the objects written to the bucket
+
+        """
+        # Determine the directory to use for file storage
+        actual_files_dir = files_dir if files_dir else tempfile.mkdtemp()
+        written_objs = []
+        # Generate random files in the specified directory
+        written_objs += generate_random_files(
+            actual_files_dir, amount, min_size, max_size
+        )
+        return written_objs
+
+    def upload_multipart_object(self, bucket_name, object_name, origin_dir):
+        """
+        Upload multipart object to the S3 bucket using boto3
+
+         Args:
+            bucket_name (str): The name of the S3 bucket.
+            object_name(str): The unique name of the S3 object
+            origin_dir (str): A directory where the files are present.
+
+        Returns:
+            Dict: Dictionary of responce generated by boto3 client
+        """
+        all_part_info = []
+        resp = self._boto3_client.create_multipart_upload(
+            Bucket=bucket_name, Key=object_name
+        )
+        file_name = origin_dir + "/" + object_name
+        upload_id = resp['UploadId']
+        part_id = 1
+        with open(file_name, "rb") as f:
+            while True:
+                log.info(f"Uploading 10 MB chunks of the file {file_name}")
+                file_chunk = f.read(10000000)
+                if file_chunk == b'':
+                    break
+                part_info = self._boto3_client.upload_part(
+                    Bucket=bucket_name,
+                    Key=object_name,
+                    PartNumber=part_id,
+                    UploadId=upload_id,
+                    Body=file_chunk
+                )
+                all_part_info.append({
+                    'PartNumber': part_id,
+                    'ETag': part_info['ETag']
+                })
+                part_id += 1
+        complete_multipart = self._boto3_client.complete_multipart_upload(
+            Bucket=bucket_name,
+            Key=object_name,
+            UploadId=upload_id,
+            MultipartUpload={
+                'Parts': all_part_info
+            },
+        )
+        return complete_multipart
```

Comment:
This method seems to upload an object in equal 10MB chunks (maybe except for the last chunk), and upload each of the parts right after their creation one after the other.

We need to separate this method into sub-parts to allow more control over the process:

We should be able to upload chunks of varrying, configurable sizes
We should be able to "pause" the process. A valid testing scenario could be to upload chunk A --> list a bucket -- > upload chunk B --> put_object on another bucket --> upload chunk C --> ... etc.
We could still use such a method / function, but it should use the sub-parts that would make the above possible.

What exceptions / errors could this method raise? We could use it for negative testing of multipart upload scenarios.

Reply:
Done